common/throttle_timer.go

```diff
@@ -12,54 +11,85 @@ If a long continuous burst of .Set() calls happens, ThrottleTimer fires
 at most once every "dur".
 */
 type ThrottleTimer struct {
-	Name string
-	Ch   chan struct{}
-	quit chan struct{}
-	dur  time.Duration
+	Name  string
+	Ch    chan struct{}
+	input chan command
+	dur   time.Duration
 
-	mtx   sync.Mutex
 	timer *time.Timer
 	isSet bool
 }
 
+type command int32
+
+const (
+	Set command = iota
+	Unset
+	Quit
+)
+
 func NewThrottleTimer(name string, dur time.Duration) *ThrottleTimer {
-	var ch = make(chan struct{})
-	var quit = make(chan struct{})
-	var t = &ThrottleTimer{Name: name, Ch: ch, dur: dur, quit: quit}
-	t.mtx.Lock()
-	t.timer = time.AfterFunc(dur, t.fireRoutine)
-	t.mtx.Unlock()
+	var t = &ThrottleTimer{
+		Name:  name,
+		Ch:    make(chan struct{}, 1),
+		dur:   dur,
+		input: make(chan command),
+		timer: time.NewTimer(dur),
+	}
 	t.timer.Stop()
+	go t.run()
 	return t
 }
 
-func (t *ThrottleTimer) fireRoutine() {
-	t.mtx.Lock()
-	defer t.mtx.Unlock()
-	select {
-	case t.Ch <- struct{}{}:
-		t.isSet = false
-	case <-t.quit:
-		// do nothing
+func (t *ThrottleTimer) run() {
+	for {
+		select {
+		case cmd := <-t.input:
+			// stop goroutine if the input says so
+			if t.processInput(cmd) {
+				// TODO: do we want to close the channels???
+				// close(t.Ch)
+				// close(t.input)
+				return
+			}
+		case <-t.timer.C:
+			t.isSet = false
+			t.Ch <- struct{}{}
```

Comment:
this was non-blocking before. I see it's buffered (cap == 1) but otherwise this is now blocking. any particular reason not to keep that old behaviour?

Reply:
In the old model, if it was not read immediately, the message got dropped, and effectively called `Set` again to schedule it for later.

This model should keep the number of ticks received equal to the number produced. The timer blocks, sure, but is there a downside to that?

Hmmm... I just checked how `time` does it, seem like they prefer to drop as you did.
https://golang.org/src/time/tick.go#L16
https://golang.org/src/time/sleep.go#L132

I'll update to that.